lib/app/preflight_check.rb

```diff
@@ -0,0 +1,87 @@
+# or more contributor license agreements. Licensed under the Elastic License;
+# you may not use this file except in compliance with the Elastic License.
+#
+
+# frozen_string_literal: true
+
+require 'app/version'
+require 'utility/logger'
+
+module App
+  class PreflightCheck
+    class CheckFailure < StandardError; end
+
+    CONNECTORS_INDEX = '.elastic-connectors'
+    JOB_INDEX = '.elastic-connectors-sync-jobs'
+
+    def self.run!
+      App::PreflightCheck.new.run!
+    end
+
+    def run!
+      check_es_connection!
+      check_es_version!
+      check_system_indices!
+    end
+
+    private
+
+    #-------------------------------------------------------------------------------------------------
+    # Checks to make sure we can connect to Elasticsearch and make API requests to it
+    def check_es_connection!
+      response = client.cluster.health
+      case response['status']
+      when 'green'
+        Utility::Logger.info('Elasticsearch backend is running and healthy.')
+      when 'yellow'
+        Utility::Logger.warn('Elasticsearch backend is running but the status is yellow.')
+      when 'red'
+        fail_check!('Elasticsearch backend is running but is unhealthy.')
+      else
+        fail_check!("Unexpected cluster status: #{response['status']}")
+      end
+    rescue StandardError
+      fail_check!('Could not connect to Elasticsearch backend. Make sure it is running and healthy.')
+    end
+
+    #-------------------------------------------------------------------------------------------------
+    # Ensures that the version of Elasticsearch is compatible with connector service
+    def check_es_version!
+      info = client.info
+      version = info.dig('version', 'number')
+      fail_check!("Cannot retrieve version from Elasticsearch response:\n#{info.to_json}") unless version
+
+      if match_es_version?(version)
+        Utility::Logger.info("Connector service version (#{App::VERSION}) matches Elasticsearch version (#{version}).")
+      else
+        fail_check!("Connector service (#{App::VERSION}) is required to run with the same major and minor version of Elasticsearch (#{version}).")
+      end
+    end
+
+    #-------------------------------------------------------------------------------------------------
+    # Ensures that the required system indices of connector service exist
+    def check_system_indices!
+      unless client.indices.exists?(:index => CONNECTORS_INDEX) && client.indices.exists?(:index => JOB_INDEX)
+        fail_check!('Required system indices for connector service don\'t exist. Make sure to run Kibana first to create system indices.')
+      end
+    end
+
+    def match_es_version?(es_version)
+      parse_minor_version(App::VERSION) == parse_minor_version(es_version)
+    end
+
+    def parse_minor_version(version)
+      version.split('.').slice(0, 2).join('.')
+    end
+
+    def client
+      @client ||= Utility::EsClient.new(App::Config[:elasticsearch])
+    end
```

Comment:
Not 100% sure, but do we want to reuse EsClient that's used by the caller?

I think in the end we'd benefit from sticking with 1 EsClient per thread (if not thread-safe) or 1 EsClient per app, which will be passed to everyone who needs it.

Reply:
In one preflight-check `run!`, the client can be reused. The `App::PreflightCheck` class is only used at the beginning of the app, i.e. one pre-flight-check per process.

>I think in the end we'd benefit from sticking with 1 EsClient per thread (if not thread-safe) or 1 EsClient per app

I think currently `EsClient` is instantiable, and instantiated by `Core::OutputSink::EsSink` and `Core::ElasticConnectorActions`. This itself is worth a discussion (see [here](https://elastic.slack.com/archives/C03GAPTQ2JF/p1662436126055189)), and a separate PR.

Reply:
Yep, agreed that it needs to be addressed as a separate PR